File: src/Framework/SPPC.Framework.Common/Presentation/FilterExpressionBuilder.cs
using System;
using System.Collections.Generic;

namespace SPPC.Framework.Presentation
{
    /// <summary>
    /// امکان ساخت یک عبارت فیلتر ساده یا ترکیبی را فراهم می کند
    /// </summary>
    public class FilterExpressionBuilder
    {
        /// <summary>
        /// نمونه جدیدی از این کلاس می سازد
        /// </summary>
        public FilterExpressionBuilder()
        {
        }

        /// <summary>
        /// عبارت فیلتر ساده یا ترکیبی ساخته شده توسط این نمونه
        /// </summary>
        public FilterExpression Expression { get; private set; }

        /// <summary>
        /// متد موزد استفاده برای شروع ساخت عبارت فیلتر
        /// </summary>
        /// <param name="filter">عبارت شرطی مورد استفاده</param>
        /// <returns>
        /// نمونه فعلی که با استفاده از آن می توان متدهای بعدی را به صورت زنجیره ای فراخوانی کرد
        /// </returns>
        public FilterExpressionBuilder New(GridFilter filter)
        {
            Expression = new FilterExpression()
            {
                Filter = filter
            };

            return this;
        }

        /// <summary>
        /// امکان ترکیب عبارت فیلتر موجود با یک عبارت شرطی جدید را به کمک عملگر منطقی "و" فراهم می کند
        /// </summary>
        /// <param name="filter">عبارت شرطی جدید</param>
        /// <returns>
        /// نمونه فعلی که با استفاده از آن می توان متدهای بعدی را به صورت زنجیره ای فراخوانی کرد
        /// </returns>
        public FilterExpressionBuilder And(GridFilter filter)
        {
            Expression.Children.Add(new FilterExpression()
            {
                Filter = filter,
                Operator = FilterExpressionOperator.And
            });

            return this;
        }

        /// <summary>
        /// امکان ترکیب عبارت فیلتر موجود با تعداد متغیری از عبارات شرطی جدید را به کمک عملگر منطقی "و" فراهم می کند
        /// </summary>
        /// <param name="filters">تعداد متغیری از عبارات شرطی جدید</param>
        /// <returns>
        /// نمونه فعلی که با استفاده از آن می توان متدهای بعدی را به صورت زنجیره ای فراخوانی کرد
        /// </returns>
        public FilterExpressionBuilder And(params GridFilter[] filters)
        {
            Array.ForEach(filters, filter => And(filter));
            return this;
        }

        /// <summary>
        /// امکان ترکیب عبارت فیلتر موجود با یک عبارت شرطی جدید را به کمک عملگر منطقی "یا" فراهم می کند
        /// </summary>
        /// <param name="filter">عبارت شرطی جدید</param>
        /// <returns>
        /// نمونه فعلی که با استفاده از آن می توان متدهای بعدی را به صورت زنجیره ای فراخوانی کرد
        /// </returns>
        public FilterExpressionBuilder Or(GridFilter filter)
        {
            Expression.Children.Add(new FilterExpression()
            {
                Filter = filter,
                Operator = FilterExpressionOperator.Or
            });

            return this;
        }

        /// <summary>
        /// امکان ترکیب عبارت فیلتر موجود با تعداد متغیری از عبارات شرطی جدید را به کمک عملگر منطقی "یا" فراهم می کند
        /// </summary>
        /// <param name="filters">تعداد متغیری از عبارات شرطی جدید</param>
        /// <returns>
        /// نمونه فعلی که با استفاده از آن می توان متدهای بعدی را به صورت زنجیره ای فراخوانی کرد
        /// </returns>
        public FilterExpressionBuilder Or(params GridFilter[] filters)
        {
            Array.ForEach(filters, filter => Or(filter));
            return this;
        }

        /// <summary>
        /// عبارت فیلتر ساخته شده را برمی گرداند
        /// </summary>
        /// <returns>عبارت فیلتر ساخته شده</returns>
        public FilterExpression Build()
        {
            return Expression;
        }
    }
}

File: src/TadbirNG/SPPC.Tadbir.Web/ClientApp/src/app/shared/class/filterExpressionBuilder.ts
import { FilterExpression } from "./filterExpression";
import { Filter } from "./filter";
import { FilterExpressionOperator } from "./filterExpressionOperator";



export class FilterExpressionBuilder{           

    public Expression: FilterExpression;

    constructor() {}

    public New(filter: Filter): FilterExpressionBuilder {
        var exp = new FilterExpression();
        exp.filter = filter;
        this.Expression = exp;
        return this;
    }

    public And(filter: Filter | Filter[]): FilterExpressionBuilder {
        if (filter instanceof Filter) {
            if (this.Expression == undefined) {
                this.New(filter);
            }
            else {
                var exp = new FilterExpression();
                exp.filter = filter;
                exp.operator = FilterExpressionOperator.And;
                //exp.parent = this.Expression;
                this.Expression.children.push(exp);
            }
        }
        else {
            for (let item of filter) {
                if (this.Expression == undefined) {
                    this.New(item);
                }
                else {
                    this.And(item);
                }
            }
        }
        return this;
    }

    public Or(filter: Filter | Filter[]): FilterExpressionBuilder {
        if (filter instanceof Filter) {
            if (this.Expression == undefined) {
                this.New(filter);
            }
            else {
                var exp = new FilterExpression();
                exp.filter = filter;
                exp.operator = FilterExpressionOperator.Or;
                //exp.parent = this.Expression;
                this.Expression.children.push(exp);
            }            
        }
        else {
            for (let item of filter) {
                if (this.Expression == undefined) {
                    this.New(item);
                }
                else {
                    this.Or(item);
                }         
            }
        }
        return this;
    }

    public Build(): FilterExpression {
        return this.Expression;
    }
}
